File: lec02/borrowing.py
import copy

h = 'hello'
h2 = h
assert(h is h2)

h = [1, 2, 3]
h2 = h
assert(h2 is h)
h2[2] = 9
assert(h[2] == 9)

h = [1, 2, 3]
h2 = h[:]
h2[2] = 9
assert(h[2] == 3)

h = [1, 2, 3]
h2 = copy.copy(h)
h2[2] = 9
assert(h[2] == 3)

h = [[1, 2, 3], [4, 5, 6]]
h2 = copy.copy(h)
h2[0][2] = 9
assert(h[0][2] == 9)

h = [[1, 2, 3], [4, 5, 6]]
h2 = copy.deepcopy(h)
h2[0][2] = 9
assert(h[1][2] == 3)

File: lec02/borrowing.rs
fn main() {
    let h = String::from("hello");
    // let h2 = append_world(&h);
    // append_world_move(h);
    let mut h2 = h.clone();
    append_world_mut(&mut h2);
    println!("{} {}", h, h2);

    let mut h = String::from("hello");

    // {
    let h1 = &h;
    let h2 = &h;
    /* let h3 = &mut h; // error here
    h3.push_str(", world"); */
    // h1 -> h?
    println!("{} {}", h1, h2);
    // }
    let h3 = &mut h;
    h3.push_str(", world");

    let _ = append_world("hello");
}

fn append_world_mut(s: &mut String) {
    s.push_str(", world");
}

// fn append_world(s: &String) -> String {
fn append_world(s: &str) -> String {
    dbg!(std::mem::size_of::<&str>());
    dbg!(std::mem::size_of::<String>());

    let s: &'static str = "hello";
    String::from("hello");

    // slice: &[T]
    // [T; N]

    /* #[derive(Clone, Copy)]
    struct SliceRepr<T> {
        ptr: *const T,
        len: usize,
    } */
    #[derive(Clone, Copy)]
    struct StrRepr {
        ptr: *const u8,
        len: usize,
    }
    union S<'a> {
        repr: StrRepr,
        st: &'a str,
    }

    unsafe {
        let repr = S { st: s }.repr;
        println!("{:?} {}", repr.ptr, repr.len);
        println!("{:?} {}", s.as_ptr(), s.len());
    }

    s.to_owned() + ", world"
    // A more efficient way to do string concatenation is by using
    // [s, ", world"].concat()
}

fn append_world_move(s: String) -> String {
    s + ", world"
}
